### PR TITLE
Fix compilation issues

### PR DESCRIPTION
**Pull request by IvanSavenko**
- gcc hardening options are now only used in debug builds - since most builders add similar (and sometimes - conflicting) flags
- fixed apparently broken compilation on FreeBSD

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -400,9 +400,11 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR NOT WIN32)
 		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-error=array-bounds") # false positives in boost::multiarray during release build, keep as warning-only
 	endif()
 
-	if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND NOT WIN32)
-		# For gcc 14+ we can use -fhardened instead
-		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -fstack-protector-strong -fstack-clash-protection -fcf-protection=full")
+	if(CMAKE_BUILD_TYPE MATCHES Debug)
```
Comment (kambala-decapitator) [inline CMakeLists.txt:403]: why not STREQUAL?

Comment (IvanSavenko) [reply CMakeLists.txt:403]: No particular reason.
Searched through file - some checks for CMAKE_BUILD_TYPE use STREQUAL, some use MATCHES.

Should I unify them to STREQUAL form?

Comment (kambala-decapitator) [reply CMakeLists.txt:403]: since we have only `Debug` config, no reason to use MATCHES for substring/wildcard search

Comment (IvanSavenko) [reply CMakeLists.txt]: Changed MATCHES to STREQUAL where applicable

**Comment by wb180** [comment]
Maybe this PR will be a good place to fix another compilation error on debug config 
OpenSUSE Tumbleweed 20240510
gcc version 13.2.1 20240206 [revision 67ac78caf31f7cb3202177e6428a46d829b70f23] 

The error is:

[1/813 0.4/sec] Building CXX object AI/FuzzyLite/fuzzylite/CMakeFiles/fl-obj.dir/src/defuzzifier/IntegralDefuzzifier.cpp.o
FAILED: AI/FuzzyLite/fuzzylite/CMakeFiles/fl-obj.dir/src/defuzzifier/IntegralDefuzzifier.cpp.o 
/usr/bin/g++ -DENABLE_EDITOR -DENABLE_LAUNCHER -DENABLE_QT_TRANSLATIONS -DFL_BACKTRACE -DFL_BUILD_PATH=\"/run/media/disk_2/vcmi/vcmi\" -DM_BIN_DIR=\"/run/media/disk_2/vcmi/vcmi/out/install/linux-gcc-debug/bin\" -DM_DATA_DIR=\"/run/media/disk_2/vcmi/vcmi/out/install/linux-gcc-debug/share/vcmi\" -DM_LIB_DIR=\"/run/media/disk_2/vcmi/vcmi/out/install/linux-gcc-debug/lib64/vcmi\" -DSCRIPTING_ENABLED=1 -DVCMI_NO_EXTRA_VERSION -DVCMI_VERSION_MAJOR=1 -DVCMI_VERSION_MINOR=5 -DVCMI_VERSION_PATCH=0 -DVCMI_VERSION_STRING=\"1.5.0\" -I/run/media/disk_2/vcmi/vcmi/AI/FuzzyLite/fuzzylite/. -pedantic -Wall -Wextra  -Wall -Wextra -Wpointer-arith -Wuninitialized -Wmismatched-tags -Wno-unused-parameter -Wno-switch -Wno-reorder -Wno-sign-compare -Wno-varargs -Werror -Wno-error=array-bounds -D_FORTIFY_SOURCE=2 -D_GLIBCXX_ASSERTIONS -fstack-protector-strong -fstack-clash-protection -fcf-protection=full -g -std=gnu++17 -fdiagnostics-color=always   -std=c++11 -fPIC -fdiagnostics-color=always -Wno-error=deprecated-declarations -MD -MT AI/FuzzyLite/fuzzylite/CMakeFiles/fl-obj.dir/src/defuzzifier/IntegralDefuzzifier.cpp.o -MF AI/FuzzyLite/fuzzylite/CMakeFiles/fl-obj.dir/src/defuzzifier/IntegralDefuzzifier.cpp.o.d -o AI/FuzzyLite/fuzzylite/CMakeFiles/fl-obj.dir/src/defuzzifier/IntegralDefuzzifier.cpp.o -c /run/media/disk_2/vcmi/vcmi/AI/FuzzyLite/fuzzylite/src/defuzzifier/IntegralDefuzzifier.cpp
In file included from /usr/include/c++/13/x86_64-suse-linux/bits/os_defines.h:39,
                 from /usr/include/c++/13/x86_64-suse-linux/bits/c++config.h:2976,
                 from /usr/include/c++/13/bits/stl_algobase.h:59,
                 from /usr/include/c++/13/algorithm:60,
                 from /run/media/disk_2/vcmi/vcmi/AI/FuzzyLite/fuzzylite/./fl/fuzzylite.h:20,
                 from /run/media/disk_2/vcmi/vcmi/AI/FuzzyLite/fuzzylite/./fl/defuzzifier/Defuzzifier.h:20,
                 from /run/media/disk_2/vcmi/vcmi/AI/FuzzyLite/fuzzylite/./fl/defuzzifier/IntegralDefuzzifier.h:20,
                 from /run/media/disk_2/vcmi/vcmi/AI/FuzzyLite/fuzzylite/src/defuzzifier/IntegralDefuzzifier.cpp:17:
/usr/include/features.h:414:4: error: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Werror=cpp]
  414 | #  warning _FORTIFY_SOURCE requires compiling with optimization (-O)
      |    ^~~~~~~
cc1plus: all warnings being treated as errors

**Pull request merged**